### PR TITLE
[Core][Refactor CoreWorker 3/n] split static function from  CoreWorkerProcess 

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

Separate the CoreWorkerProcess static functions from CoreWorkerProcess state; Currently the static and non-static state are mixed together, and more importantly the static state is not thread safe. By separating them and create helper class for non-static state `CoreWorkerProcessImpl`, we can make it thread safe.

in follow up PR we will make CoreWorkerProcess state thread safe. 

This PR depends on #19677,  The follow up PR is  #19679

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker_process.cc [hunk 1]
```diff
@@ -21,36 +21,66 @@ namespace ray {
...
-thread_local std::weak_ptr<CoreWorker> CoreWorkerProcess::current_core_worker_;
+}  // namespace
 
 void CoreWorkerProcess::Initialize(const CoreWorkerOptions &options) {
   RAY_CHECK(!core_worker_process)
       << "The process is already initialized for core worker.";
-  core_worker_process.reset(new CoreWorkerProcess(options));
+  core_worker_process.reset(new CoreWorkerProcessImpl(options));
+
+#ifndef _WIN32
```
Comment (scv119) [inline src/ray/core_worker/core_worker_process.cc:33]: @kfstorm  this is the only semantical change: we moved `HandleAtExit` registration to `CoreWorkerProcess::Initialize`

Comment (kfstorm) [reply src/ray/core_worker/core_worker_process.cc:33]: Is there any particular reason to do this?

Comment (scv119) [reply src/ray/core_worker/core_worker_process.cc:33]: CoreWorkerProcess is now a static class and doesn't have constructor; also we can put it into CoreWorkerProcessImpl as it doesn't/shouldn't have access to the static class info.

File: src/ray/core_worker/core_worker_process.h [hunk 2]
```diff
@@ -151,16 +152,28 @@ class CoreWorkerProcess {
...
   const CoreWorkerOptions options_;
 
-  /// The core worker instance associated with the current thread.
-  /// Use weak_ptr here to avoid memory leak due to multi-threading.
-  static thread_local std::weak_ptr<CoreWorker> current_core_worker_;
-
   /// The only core worker instance, if the number of workers is 1.
   std::shared_ptr<CoreWorker> global_worker_ GUARDED_BY(mutex_);
 
+  static thread_local std::weak_ptr<CoreWorker> thread_local_core_worker_;
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker_process.h:175]: We can keep the docstring?

**Review by rkooo567 — APPROVED** [review]
LGTM given most of changes are from moving pieces

**Review by kfstorm — APPROVED** [review]
LGTM overall.

File: src/ray/core_worker/core_worker_process.cc (hunk 1 above)
Comment (kfstorm) [reply src/ray/core_worker/core_worker_process.cc:33]: I see.

**Comment by scv119** [comment]
test_basic_2_client is flaky on master.

**Pull request merged**